#### Converting the table from the Wikipedia page in a Pandas dataframe.

In [39]:
import pandas as pd
link = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
table = pd.read_html(link)[0]

In [40]:
table.head(15)

Postcode           Borough     Neighbourhood
0       M1A      Not assigned      Not assigned
1       M2A      Not assigned      Not assigned
2       M3A        North York         Parkwoods
3       M4A        North York  Victoria Village
4       M5A  Downtown Toronto      Harbourfront
5       M6A        North York  Lawrence Heights
6       M6A        North York    Lawrence Manor
7       M7A  Downtown Toronto      Queen's Park
8       M8A      Not assigned      Not assigned
9       M9A      Queen's Park      Not assigned
10      M1B       Scarborough             Rouge
11      M1B       Scarborough           Malvern
12      M2B      Not assigned      Not assigned
13      M3B        North York   Don Mills North
14      M4B         East York  Woodbine Gardens

In [41]:
no_borough = table[table["Borough"]=="Not assigned"]
table.drop(no_borough.index,inplace = True)
table.reset_index(drop=True)

Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
3        M6A        North York   
4        M6A        North York   
5        M7A  Downtown Toronto   
6        M9A      Queen's Park   
7        M1B       Scarborough   
8        M1B       Scarborough   
9        M3B        North York   
10       M4B         East York   
11       M4B         East York   
12       M5B  Downtown Toronto   
13       M5B  Downtown Toronto   
14       M6B        North York   
15       M9B         Etobicoke   
16       M9B         Etobicoke   
17       M9B         Etobicoke   
18       M9B         Etobicoke   
19       M9B         Etobicoke   
20       M1C       Scarborough   
21       M1C       Scarborough   
22       M1C       Scarborough   
23       M3C        North York   
24       M3C        North York   
25       M4C         East York   
26       M5C  Downtown Toronto   
27       M6C              York   
28       M9C         Etobicoke   
29       M9C         Etobicoke   
30       M9C         Etobicoke   
31       M9C         Etobicoke   
32       M1E       Scarborough   
33       M1E       Scarborough   
34       M1E       Scarborough   
35       M4E      East Toronto   
36       M5E  Downtown Toronto   
37       M6E              York   
38       M1G       Scarborough   
39       M4G         East York   
40       M5G  Downtown Toronto   
41       M6G  Downtown Toronto   
42       M1H       Scarborough   
43       M2H        North York   
44       M3H        North York   
45       M3H        North York   
46       M3H        North York   
47       M4H         East York   
48       M5H  Downtown Toronto   
49       M5H  Downtown Toronto   
50       M5H  Downtown Toronto   
51       M6H      West Toronto   
52       M6H      West Toronto   
53       M1J       Scarborough   
54       M2J        North York   
55       M2J        North York   
56       M2J        North York   
57       M3J        North York   
58       M3J        North York   
59       M4J         East York   
60       M5J  Downtown Toronto   
61       M5J  Downtown Toronto   
62       M5J  Downtown Toronto   
63       M6J      West Toronto   
64       M6J      West Toronto   
65       M1K       Scarborough   
66       M1K       Scarborough   
67       M1K       Scarborough   
68       M2K        North York   
69       M3K        North York   
70       M3K        North York   
71       M4K      East Toronto   
72       M4K      East Toronto   
73       M5K  Downtown Toronto   
74       M5K  Downtown Toronto   
75       M6K      West Toronto   
76       M6K      West Toronto   
77       M6K      West Toronto   
78       M1L       Scarborough   
79       M1L       Scarborough   
80       M1L       Scarborough   
81       M2L        North York   
82       M2L        North York   
83       M3L        North York   
84       M4L      East Toronto   
85       M4L      East Toronto   
86       M5L  Downtown Toronto   
87       M5L  Downtown Toronto   
88       M6L        North York   
89       M6L        North York   
90       M6L        North York   
91       M9L        North York   
92       M1M       Scarborough   
93       M1M       Scarborough   
94       M1M       Scarborough   
95       M2M        North York   
96       M2M        North York   
97       M3M        North York   
98       M4M      East Toronto   
99       M5M        North York   
100      M5M        North York   
101      M6M              York   
102      M6M              York   
103      M6M              York   
104      M6M              York   
105      M9M        North York   
106      M9M        North York   
107      M1N       Scarborough   
108      M1N       Scarborough   
109      M2N        North York   
110      M3N        North York   
111      M4N   Central Toronto   
112      M5N   Central Toronto   
113      M6N              York   
114      M6N              York   
115      M9N              York   
116      M1P       Scarbor

In [42]:
s = table.shape
for i in range(0,s[0]):
    if table["Neighbourhood"].iloc[i]=="Not assigned":
        table.replace("Not assigned",table["Borough"].iloc[i],inplace = True)
        

In [43]:
table.reset_index(drop=True,inplace = True)

In [44]:
table.head(20)

Postcode           Borough     Neighbourhood
0       M3A        North York         Parkwoods
1       M4A        North York  Victoria Village
2       M5A  Downtown Toronto      Harbourfront
3       M6A        North York  Lawrence Heights
4       M6A        North York    Lawrence Manor
5       M7A  Downtown Toronto      Queen's Park
6       M9A      Queen's Park      Queen's Park
7       M1B       Scarborough             Rouge
8       M1B       Scarborough           Malvern
9       M3B        North York   Don Mills North
10      M4B         East York  Woodbine Gardens
11      M4B         East York     Parkview Hill
12      M5B  Downtown Toronto           Ryerson
13      M5B  Downtown Toronto   Garden District
14      M6B        North York         Glencairn
15      M9B         Etobicoke        Cloverdale
16      M9B         Etobicoke         Islington
17      M9B         Etobicoke      Martin Grove
18      M9B         Etobicoke  Princess Gardens
19      M9B         Etobicoke   West Deane Park

In [45]:
table.shape

(210, 3)

In [46]:
table_new = table.groupby(["Postcode","Borough"],axis = 0)["Neighbourhood"].apply(lambda x: ','.join(x))
table_new  = pd.DataFrame(table_new)
table_new.reset_index(drop=False,inplace = True)

In [47]:
table_new.head(60)

Postcode           Borough  \
0       M1B       Scarborough   
1       M1C       Scarborough   
2       M1E       Scarborough   
3       M1G       Scarborough   
4       M1H       Scarborough   
5       M1J       Scarborough   
6       M1K       Scarborough   
7       M1L       Scarborough   
8       M1M       Scarborough   
9       M1N       Scarborough   
10      M1P       Scarborough   
11      M1R       Scarborough   
12      M1S       Scarborough   
13      M1T       Scarborough   
14      M1V       Scarborough   
15      M1W       Scarborough   
16      M1X       Scarborough   
17      M2H        North York   
18      M2J        North York   
19      M2K        North York   
20      M2L        North York   
21      M2M        North York   
22      M2N        North York   
23      M2P        North York   
24      M2R        North York   
25      M3A        North York   
26      M3B        North York   
27      M3C        North York   
28      M3H        North York   
29      M3J        North York   
30      M3K        North York   
31      M3L        North York   
32      M3M        North York   
33      M3N        North York   
34      M4A        North York   
35      M4B         East York   
36      M4C         East York   
37      M4E      East Toronto   
38      M4G         East York   
39      M4H         East York   
40      M4J         East York   
41      M4K      East Toronto   
42      M4L      East Toronto   
43      M4M      East Toronto   
44      M4N   Central Toronto   
45      M4P   Central Toronto   
46      M4R   Central Toronto   
47      M4S   Central Toronto   
48      M4T   Central Toronto   
49      M4V   Central Toronto   
50      M4W  Downtown Toronto   
51      M4X  Downtown Toronto   
52      M4Y  Downtown Toronto   
53      M5A  Downtown Toronto   
54      M5B  Downtown Toronto   
55      M5C  Downtown Toronto   
56      M5E  Downtown Toronto   
57      M5G  Downtown Toronto   
58      M5H  Downtown Toronto   
59      M5J  Downtown Toronto   

                                        Neighbourhood  
0                                       Rouge,Malvern  
1                Highland Creek,Rouge Hill,Port Union  
2                     Guildwood,Morningside,West Hill  
3                                              Woburn  
4                                           Cedarbrae  
5                                 Scarborough Village  
6           East Birchmount Park,Ionview,Kennedy Park  
7                       Clairlea,Golden Mile,Oakridge  
8       Cliffcrest,Cliffside,Scarborough Village West  
9                          Birch Cliff,Cliffside West  
10  Dorset Park,Scarborough Town Centre,Wexford He...  
11                                   Maryvale,Wexford  
12                                          Agincourt  
13              Clarks Corners,Sullivan,Tam O'Shanter  
14  Agincourt North,L'Amoreaux East,Milliken,Steel...  
15                                    L'Amoreaux West  
16                                        Upper Rouge  
17                                  Hillcrest Village  
18                         Fairview,Henry Farm,Oriole  
19                                    Bayview Village  
20                            Silver Hills,York Mills  
21                             Newtonbrook,Willowdale  
22                                   Willowdale South  
23                                    York Mills West  
24                                    Willowdale West  
25                                          Parkwoods  
26                                    Don Mills North  
27                    Flemingdon Park,Don Mills South  
28      Bathurst Manor,Downsview North,Wilson Heights  
29                     Northwood Park,York University  
30                         CFB Toronto,Downsview East  
31                                     Downsview West  
32                                  Downsview Central  
33                                Downsview Northwest  
34                           

#### Importing required libraries.

In [48]:
import numpy as np
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
import json 
from geopy.geocoders import Nominatim
import requests
from pandas.io.json import json_normalize
import matplotlib.cm as cm
import matplotlib.colors as colors
from sklearn.cluster import KMeans
import folium

#### Reading the location CSV file-

In [49]:
loc_data = pd.read_csv("Geospatial_Coordinates.csv")

In [50]:
loc_data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
5           M1J  43.744734 -79.239476
6           M1K  43.727929 -79.262029
7           M1L  43.711112 -79.284577
8           M1M  43.716316 -79.239476
9           M1N  43.692657 -79.264848
10          M1P  43.757410 -79.273304
11          M1R  43.750072 -79.295849
12          M1S  43.794200 -79.262029
13          M1T  43.781638 -79.304302
14          M1V  43.815252 -79.284577
15          M1W  43.799525 -79.318389
16          M1X  43.836125 -79.205636
17          M2H  43.803762 -79.363452
18          M2J  43.778517 -79.346556
19          M2K  43.786947 -79.385975
20          M2L  43.757490 -79.374714
21          M2M  43.789053 -79.408493
22          M2N  43.770120 -79.408493
23          M2P  43.752758 -79.400049
24          M2R  43.782736 -79.442259
25          M3A  43.753259 -79.329656
26          M3B  43.745906 -79.352188
27          M3C  43.725900 -79.340923
28          M3H  43.754328 -79.442259
29          M3J  43.767980 -79.487262
30          M3K  43.737473 -79.464763
31          M3L  43.739015 -79.506944
32          M3M  43.728496 -79.495697
33          M3N  43.761631 -79.520999
34          M4A  43.725882 -79.315572
35          M4B  43.706397 -79.309937
36          M4C  43.695344 -79.318389
37          M4E  43.676357 -79.293031
38          M4G  43.709060 -79.363452
39          M4H  43.705369 -79.349372
40          M4J  43.685347 -79.338106
41          M4K  43.679557 -79.352188
42          M4L  43.668999 -79.315572
43          M4M  43.659526 -79.340923
44          M4N  43.728020 -79.388790
45          M4P  43.712751 -79.390197
46          M4R  43.715383 -79.405678
47          M4S  43.704324 -79.388790
48          M4T  43.689574 -79.383160
49          M4V  43.686412 -79.400049
50          M4W  43.679563 -79.377529
51          M4X  43.667967 -79.367675
52          M4Y  43.665860 -79.383160
53          M5A  43.654260 -79.360636
54          M5B  43.657162 -79.378937
55          M5C  43.651494 -79.375418
56          M5E  43.644771 -79.373306
57          M5G  43.657952 -79.387383
58          M5H  43.650571 -79.384568
59          M5J  43.640816 -79.381752
60          M5K  43.647177 -79.381576
61          M5L  43.648198 -79.379817
62          M5M  43.733283 -79.419750
63          M5N  43.711695 -79.416936
64          M5P  43.696948 -79.411307
65          M5R  43.672710 -79.405678
66          M5S  43.662696 -79.400049
67          M5T  43.653206 -79.400049
68          M5V  43.628947 -79.394420
69          M5W  43.646435 -79.374846
70          M5X  43.648429 -79.382280
71          M6A  43.718518 -79.464763
72          M6B  43.709577 -79.445073
73          M6C  43.693781 -79.428191
74          M6E  43.689026 -79.453512
75          M6G  43.669542 -79.422564
76          M6H  43.669005 -79.442259
77          M6J  43.647927 -79.419750
78          M6K  43.636847 -79.428191
79          M6L  43.713756 -79.490074
80          M6M  43.691116 -79.476013
81          M6N  43.673185 -79.487262
82          M6P  43.661608 -79.464763
83          M6R  43.648960 -79.456325
84          M6S  43.651571 -79.484450
85          M7A  43.662301 -79.389494
86          M7R  43.636966 -79.615819
87          M7Y  43.662744 -79.321558
88          M8V  43.605647 -79.501321
89          M8W  43.602414 -79.543484
90          M8X  43.653654 -79.506944
91          M8Y  43.636258 -79.498509
92          M8Z  43.628841 -79.520999
93          M9A  43.667856 -79.532242
94          M9B  43.650943 -79.554724
95          M9C  43.643515 -79.577201
96          M9L  43.756303 -79.565963
97          M9M  43.724766 -79.532242
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

#### As we can see, both of the daraframes has the same number of rows,i.e. 103.

#### Renaming the column "Postal code" as "Postcode" so that we can merge the dataframes.

In [51]:
loc_data.rename(columns={"Postal Code":"Postcode"},inplace = True)

In [52]:
loc_data.head()

Postcode   Latitude  Longitude
0      M1B  43.806686 -79.194353
1      M1C  43.784535 -79.160497
2      M1E  43.763573 -79.188711
3      M1G  43.770992 -79.216917
4      M1H  43.773136 -79.239476

In [53]:
loc_data.shape

(103, 3)

#### Merging both the dataframes to get the desired dataframe which has "Postcode","Borough","Neighbourhood","Latitude" and "Longitude".

In [54]:
table_with_location = pd.merge(table_new,loc_data,how = "outer")
table_with_location.rename(columns = {"Postcode":"PostalCode"},inplace = True)
table_with_location.head()

PostalCode      Borough                         Neighbourhood   Latitude  \
0        M1B  Scarborough                         Rouge,Malvern  43.806686   
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union  43.784535   
2        M1E  Scarborough       Guildwood,Morningside,West Hill  43.763573   
3        M1G  Scarborough                                Woburn  43.770992   
4        M1H  Scarborough                             Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

#### We need the location data of Toronto to segment and cluster the neighbourhood.

In [65]:
table_borough = table_with_location[["Borough"]]
toronto_table = table_with_location[table_borough.apply(lambda row: row.astype(str).str.contains('Toronto').any(), axis=1)]
toronto_table.reset_index(drop = True,inplace = True)
toronto_table.head()

PostalCode          Borough                  Neighbourhood   Latitude  \
0        M4E     East Toronto                    The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West,Riverdale  43.679557   
2        M4L     East Toronto  The Beaches West,India Bazaar  43.668999   
3        M4M     East Toronto                Studio District  43.659526   
4        M4N  Central Toronto                  Lawrence Park  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

#### Getting the Latitude and Longitude of Toronto, Canada - 

In [57]:
address = 'Toronto, CANADA'

geolocator = Nominatim(user_agent="Toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


#### Creating a map of Toronto using Folium library

In [59]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

#### Superimposing the neighbourhoods obtained from the dataframe <b>"toronto_table"</b> on the <b>"map_toronto"</b>-

In [60]:
for lat, lng, borough, neighbourhood in zip(toronto_table['Latitude'], toronto_table['Longitude'], toronto_table['Borough'], toronto_table['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='green',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

#### To cluster the neighbourhoods in Toronto, we need Foursquare API client ID and client Secret.

In [64]:
{
    "tags": [
        "remove_cell",
    ]
}



CLIENT_ID = 'P2LTGBSPUIWKRMZF55OPZEW0H33JE45DQ225YFEVQWPRKVNV'
CLIENT_SECRET = 'T3NST1UUDSME3VDVIXGINPWOXHWY0VELWYXGGUFCAUKX3DY5'
VERSION = '20180605'

In [66]:
toronto_table.loc[0,"Neighbourhood"]

'The Beaches'

In [67]:
neighbourhood_latitude = toronto_table.loc[0, 'Latitude'] # neighborhood latitude value
neighbourhood_longitude = toronto_table.loc[0, 'Longitude'] # neighborhood longitude value

neighbourhood_name = toronto_table.loc[0, 'Neighbourhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighbourhood_name, 
                                                               neighbourhood_latitude, 
                                                               neighbourhood_longitude))

Latitude and longitude values of The Beaches are 43.67635739999999, -79.2930312.


In [68]:
LIMIT  = 100
radius = 500


url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighbourhood_latitude, 
    neighbourhood_longitude, 
    radius, 
    LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?&client_id=P2LTGBSPUIWKRMZF55OPZEW0H33JE45DQ225YFEVQWPRKVNV&client_secret=T3NST1UUDSME3VDVIXGINPWOXHWY0VELWYXGGUFCAUKX3DY5&v=20180605&ll=43.67635739999999,-79.2930312&radius=500&limit=100'

In [69]:
results = requests.get(url).json()
results 

{'meta': {'code': 200, 'requestId': '5e4a5ff11835dd001bc3c0ad'},
 'response': {'headerLocation': 'The Beaches',
  'headerFullLocation': 'The Beaches, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 5,
  'suggestedBounds': {'ne': {'lat': 43.680857404499996,
    'lng': -79.28682091449052},
   'sw': {'lat': 43.67185739549999, 'lng': -79.29924148550948}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4bd461bc77b29c74a07d9282',
       'name': 'Glen Manor Ravine',
       'location': {'address': 'Glen Manor',
        'crossStreet': 'Queen St.',
        'lat': 43.67682094413784,
        'lng': -79.29394208780985,
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.67682094413784,
          'lng': -79.29394208780985}],
        'distanc

In [70]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [71]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues)


filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]


nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)


nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head(5)

/home/debaparna/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  This is separate from the ipykernel package so we can avoid doing imports until


name         categories        lat        lng
0                   Glen Manor Ravine              Trail  43.676821 -79.293942
1  The Big Carrot Natural Food Market  Health Food Store  43.678879 -79.297734
2                 Grover Pub and Grub                Pub  43.679181 -79.297215
3                   Glen Stewart Park               Park  43.675278 -79.294647
4                       Upper Beaches       Neighborhood  43.680563 -79.292869

In [ ]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

#### getNearbyVenues by Foursquare-

In [73]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [75]:
toronto_venues = getNearbyVenues(names=toronto_table['Neighbourhood'],
                                   latitudes=toronto_table['Latitude'],
                                   longitudes=toronto_table['Longitude']
                                  )

The Beaches
The Danforth West,Riverdale
The Beaches West,India Bazaar
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park,Summerhill East
Deer Park,Forest Hill SE,Rathnelly,South Hill,Summerhill West
Rosedale
Cabbagetown,St. James Town
Church and Wellesley
Harbourfront
Ryerson,Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide,King,Richmond
Harbourfront East,Toronto Islands,Union Station
Design Exchange,Toronto Dominion Centre
Commerce Court,Victoria Hotel
Roselawn
Forest Hill North,Forest Hill West
The Annex,North Midtown,Yorkville
Harbord,University of Toronto
Chinatown,Grange Park,Kensington Market
CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place,Underground city
Christie
Dovercourt Village,Dufferin
Little Portugal,Trinity
Brockton,Exhibition Place,Parkdale Village
High Park,The Junction South
Parkdale,Roncesvalles
Runnymede

#### Shape of the dataframe

In [76]:
toronto_venues.shape


(1701, 7)

#### The dataframe looks like this-

In [77]:
toronto_venues.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0   The Beaches               43.676357               -79.293031   
1   The Beaches               43.676357               -79.293031   
2   The Beaches               43.676357               -79.293031   
3   The Beaches               43.676357               -79.293031   
4   The Beaches               43.676357               -79.293031   

                                Venue  Venue Latitude  Venue Longitude  \
0                   Glen Manor Ravine       43.676821       -79.293942   
1  The Big Carrot Natural Food Market       43.678879       -79.297734   
2                 Grover Pub and Grub       43.679181       -79.297215   
3                   Glen Stewart Park       43.675278       -79.294647   
4                       Upper Beaches       43.680563       -79.292869   

      Venue Category  
0              Trail  
1  Health Food Store  
2                Pub  
3               Park  
4       Neighborhood

#### checking venues for each neighbourhood-

In [78]:
toronto_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                                
Adelaide,King,Richmond                                                 100   
Berczy Park                                                             56   
Brockton,Exhibition Place,Parkdale Village                              25   
Business Reply Mail Processing Centre 969 Eastern                       18   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                      16   
Cabbagetown,St. James Town                                              45   
Central Bay Street                                                      83   
Chinatown,Grange Park,Kensington Market                                 80   
Christie                                                                18   
Church and Wellesley                                                    84   
Commerce Court,Victoria Hotel                                          100   
Davisville                                                              35   
Davisville North                                                         8   
Deer Park,Forest Hill SE,Rathnelly,South Hill,S...                      14   
Design Exchange,Toronto Dominion Centre                                100   
Dovercourt Village,Dufferin                                             15   
First Canadian Place,Underground city                                  100   
Forest Hill North,Forest Hill West                                       4   
Harbord,University of Toronto                                           35   
Harbourfront                                                            46   
Harbourfront East,Toronto Islands,Union Station                        100   
High Park,The Junction South                                            23   
Lawrence Park                                                            3   
Little Portugal,Trinity                                                 53   
Moore Park,Summerhill East                                               2   
North Toronto West                                                      19   
Parkdale,Roncesvalles                                                   14   
Queen's Park                                                            39   
Rosedale                                                                 4   
Roselawn                                                                 1   
Runnymede,Swansea                                                       36   
Ryerson,Garden District                                                100   
St. James Town                                                         100   
Stn A PO Boxes 25 The Esplanade                                         95   
Studio District                                                         41   
The Annex,North Midtown,Yorkville                                       22   
The Beaches                                                              5   
The Beaches West,India Bazaar                                           22   
The Danforth West,Riverdale                                             40   

                                                    Neighbourhood Longitude  \
Neighbourhood                                                                 
Adelaide,King,Richmond                                                  100   
Berczy Park                                                              56   
Brockton,Exhibition Place,Parkdale Village                               25   
Business Reply Mail Processing Centre 969 Eastern                        18   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                       16   
Cabbagetown,St. James Town                                               45   
Central Bay Street                                                       83   
Chinatown,Grange Park,Kensington Market                                  80   
Christie                                                       

#### Checking unique catagories-

In [79]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 227 uniques categories.


In [80]:
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighbourhood  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0   The Beaches                  0        0                   0             0   
1   The Beaches                  0        0                   0             0   
2   The Beaches                  0        0                   0             0   
3   The Beaches                  0        0                   0             0   
4   The Beaches                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Arts & Crafts Store  Asian Restaurant  \
0             0         0            0                    0                 0   
1             0         0            0                    0                 0   
2             0         0            0                    0                 0   
3             0         0            0                    0                 0   
4             0         0            0                    0                 0   

   Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0                   0              0          0           0           0   
1                   0              0          0           0           0   
2                   0              0          0           0           0   
3                   0              0          0           0           0   
4                   0              0          0           0           0   

   Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0       0     0    0                 0                   0      0   
1       0     0    0                 0                   0      0   
2       0     0    0                 0                   0      0   
3       0     0    0                 0                   0      0   
4       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                0         0           0                   0       0   
1                0         0           0                   0       0   
2                0         0           0                   0       0   
3                0         0           0                   0       0   
4                0         0           0                   0       0   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0              0          0         0                     0               0   
1              0          0         0                     0               0   
2              0          0         0                     0               0   
3              0          0         0                     0               0   
4              0          0         0                     0               0   

   Brewery  Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0        0                0         0             0              0         0   
1        0                0         0             0              0         0   
2        0                0         0             0              0         0   
3        0                0         0             0              0         0   
4        0                0         0             0              0         0   

   Butcher  Café  Cajun / Creole Restaurant  Camera Store  Candy Store  \
0        0     0                          0             0            0   
1        0     0                          0             0            0   
2        0     0                          0             0            0   
3        0    

In [81]:
toronto_onehot.shape

(1701, 228)

In [82]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Neighbourhood  Afghan Restaurant  \
0                              Adelaide,King,Richmond           0.000000   
1                                         Berczy Park           0.000000   
2          Brockton,Exhibition Place,Parkdale Village           0.000000   
3   Business Reply Mail Processing Centre 969 Eastern           0.000000   
4   CN Tower,Bathurst Quay,Island airport,Harbourf...           0.000000   
5                          Cabbagetown,St. James Town           0.000000   
6                                  Central Bay Street           0.000000   
7             Chinatown,Grange Park,Kensington Market           0.000000   
8                                            Christie           0.000000   
9                                Church and Wellesley           0.011905   
10                      Commerce Court,Victoria Hotel           0.000000   
11                                         Davisville           0.000000   
12                                   Davisville North           0.000000   
13  Deer Park,Forest Hill SE,Rathnelly,South Hill,...           0.000000   
14            Design Exchange,Toronto Dominion Centre           0.000000   
15                        Dovercourt Village,Dufferin           0.000000   
16              First Canadian Place,Underground city           0.000000   
17                 Forest Hill North,Forest Hill West           0.000000   
18                      Harbord,University of Toronto           0.000000   
19                                       Harbourfront           0.000000   
20    Harbourfront East,Toronto Islands,Union Station           0.000000   
21                       High Park,The Junction South           0.000000   
22                                      Lawrence Park           0.000000   
23                            Little Portugal,Trinity           0.000000   
24                         Moore Park,Summerhill East           0.000000   
25                                 North Toronto West           0.000000   
26                              Parkdale,Roncesvalles           0.000000   
27                                       Queen's Park           0.000000   
28                                           Rosedale           0.000000   
29                                           Roselawn           0.000000   
30                                  Runnymede,Swansea           0.000000   
31                            Ryerson,Garden District           0.000000   
32                                     St. James Town           0.000000   
33                    Stn A PO Boxes 25 The Esplanade           0.000000   
34                                    Studio District           0.000000   
35                  The Annex,North Midtown,Yorkville           0.000000   
36                                        The Beaches           0.000000   
37                      The Beaches West,India Bazaar           0.000000   
38                        The Danforth West,Riverdale           0.000000   

    Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0    0.0000              0.0000        0.0000           0.000   
1    0.0000              0.0000        0.0000           0.000   
2    0.0000              0.0000        0.0000           0.000   
3    0.0000              0.0000        0.0000           0.000   
4    0.0625              0.0625        0.0625           0.125   
5    0.0000              0.0000        0.0000           0.000   
6    0.0000              0.0000        0.0000           0.000   
7    0.0000              0.0000        0.0000           0.000   
8    0.0000              0.0000        0.0000           0.000   
9    0.0000              0.0000        0.0000           0.000   
10   0.0000              0.0000        0.0000           0.000   
11   0.0000              0.0000        0.0000           0.000   
12   0.0000              0.0000        0.0000           0.000   
13   0.0000              0.0000        0.0000           0.000   
14   0.0000             

In [84]:
toronto_grouped.shape

(39, 228)

#### Each neighborhood along with the top 5 most common venues-

In [85]:
num_top_venues = 5

for hood in toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide,King,Richmond----
              venue  freq
0       Coffee Shop  0.07
1              Café  0.04
2        Steakhouse  0.04
3   Thai Restaurant  0.04
4  Asian Restaurant  0.03


----Berczy Park----
                venue  freq
0         Coffee Shop  0.09
1        Cocktail Bar  0.04
2  Seafood Restaurant  0.04
3         Cheese Shop  0.04
4      Farmers Market  0.04


----Brockton,Exhibition Place,Parkdale Village----
                   venue  freq
0                   Café  0.12
1         Breakfast Spot  0.08
2  Performing Arts Venue  0.08
3              Nightclub  0.08
4            Coffee Shop  0.08


----Business Reply Mail Processing Centre 969 Eastern----
                venue  freq
0  Light Rail Station  0.11
1         Yoga Studio  0.06
2       Garden Center  0.06
3         Pizza Place  0.06
4          Comic Shop  0.06


----CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara----
              venue  freq
0    Airport Lounge

In [86]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [88]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head(10)

Neighbourhood 1st Most Common Venue  \
0                             Adelaide,King,Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2         Brockton,Exhibition Place,Parkdale Village                  Café   
3  Business Reply Mail Processing Centre 969 Eastern    Light Rail Station   
4  CN Tower,Bathurst Quay,Island airport,Harbourf...        Airport Lounge   
5                         Cabbagetown,St. James Town           Coffee Shop   
6                                 Central Bay Street           Coffee Shop   
7            Chinatown,Grange Park,Kensington Market                   Bar   
8                                           Christie         Grocery Store   
9                               Church and Wellesley           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0             Steakhouse       Thai Restaurant                  Café   
1                   Café           Cheese Shop        Farmers Market   
2  Performing Arts Venue        Breakfast Spot             Nightclub   
3            Yoga Studio           Pizza Place            Skate Park   
4       Airport Terminal           Coffee Shop              Boutique   
5            Pizza Place    Italian Restaurant                   Pub   
6     Italian Restaurant        Sandwich Place             Juice Bar   
7                   Café    Chinese Restaurant           Coffee Shop   
8                   Café                  Park            Restaurant   
9    Japanese Restaurant               Gay Bar      Sushi Restaurant   

  5th Most Common Venue  6th Most Common Venue          7th Most Common Venue  \
0          Burger Joint                    Bar                         Bakery   
1                Bakery     Seafood Restaurant                     Steakhouse   
2           Coffee Shop            Yoga Studio                            Gym   
3         Burrito Place                    Spa                 Farmers Market   
4                   Bar    Rental Car Location               Sculpture Garden   
5                Bakery             Restaurant                           Café   
6   Japanese Restaurant           Burger Joint                 Ice Cream Shop   
7   Dumpling Restaurant  Vietnamese Restaurant  Vegetarian / Vegan Restaurant   
8            Baby Store     Italian Restaurant             Athletics & Sports   
9            Restaurant                   Café                            Gym   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0      Sushi Restaurant      Asian Restaurant         Cosmetics Shop  
1              Beer Bar          Cocktail Bar                 Lounge  
2             Pet Store    Italian Restaurant           Intersection  
3  Fast Food Restaurant               Butcher             Restaurant  
4                 Plane         Boat or Ferry        Harbor / Marina  
5        Ice Cream Shop     Indian Restaurant         Breakfast Spot  
6                  Café           Salad Place     Chinese Restaurant  
7    Mexican Restaurant         Grocery Store                   Park  
8                 Diner             Nightclub            Coffee Shop  
9                 Hotel             Gastropub   Fast Food Restaurant

#### Now we've got the popular places so that we can cluster the neighbourhoods.

#### Running K-means clustering method for k = 5 clusters.

In [89]:
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

In [96]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_table

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head() # check the last columns!

PostalCode          Borough                  Neighbourhood   Latitude  \
0        M4E     East Toronto                    The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West,Riverdale  43.679557   
2        M4L     East Toronto  The Beaches West,India Bazaar  43.668999   
3        M4M     East Toronto                Studio District  43.659526   
4        M4N  Central Toronto                  Lawrence Park  43.728020   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.293031               0                 Trail     Health Food Store   
1 -79.352188               0      Greek Restaurant    Italian Restaurant   
2 -79.315572               0                  Park        Sandwich Place   
3 -79.340923               0                  Café           Coffee Shop   
4 -79.388790               0                  Park           Swim School   

  3rd Most Common Venue 4th Most Common Venue   5th Most Common Venue  \
0                  Park                   Pub            Neighborhood   
1           Coffee Shop        Ice Cream Shop  Furniture / Home Store   
2     Food & Drink Shop          Burger Joint           Burrito Place   
3             Gastropub               Brewery                  Bakery   
4              Bus Line      Department Store       Electronics Store   

         6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                 Dessert Shop    Dim Sum Restaurant                 Diner   
1                  Yoga Studio    Frozen Yogurt Shop                   Pub   
2           Italian Restaurant  Fast Food Restaurant     Fish & Chips Shop   
3           Italian Restaurant   American Restaurant        Sandwich Place   
4  Eastern European Restaurant   Dumpling Restaurant            Donut Shop   

  9th Most Common Venue 10th Most Common Venue  
0        Discount Store            Yoga Studio  
1           Pizza Place           Liquor Store  
2            Steakhouse         Ice Cream Shop  
3           Cheese Shop              Pet Store  
4      Doner Restaurant                Dog Run

#### Adding clusters to the Toronto map-

In [100]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters